lib/spack/spack/build_systems/oneapi.py

```diff
@@ -106,6 +106,10 @@ def setup_run_environment(self, env):
         env.extend(EnvironmentModifications.from_sourcing_file(
             join_path(self.component_path, 'env', 'vars.sh')))
 
+    def setup_dependent_build_environment(self, env, dependent_spec):
+        env.extend(EnvironmentModifications.from_sourcing_file(
+            join_path(self.component_path, 'env', 'vars.sh')))
+
```

Comment:
If I understand correctly:

* `vars.sh` sets `MKLROOT`
* Therefore, all `IntelOneApiPackage`s now set `MKLROOT`

I'm concerned if this is true for other important env varables, and whether that might create a conflict when there is a mixture of oneAPI and non-oneAPI dependencies.

Two notes:

* This is less of a concern for `MKLROOT` specifically because only Intel packages ever provide `mkl`, and it's unlikely there would be a mixture of oneAPI. I'm more concerned if it sets any "generic" variables (e.g. related to `mpi`)
* It's generally less problematic for `setup_run_environment` to do this (I can provide more details on that if desired but I wanted to clarify that since it might seem contradictory given that we already do this for `setup_run_environment`)

Also, I'd like to clarify a couple things about the user's issue in https://github.com/spack/spack/discussions/30747#discussioncomment-2859855 before reviewing this more.

Reply:
> * Therefore, all `IntelOneApiPackage`s now set `MKLROOT`

This is not the way it works. Each package has its own vars.sh.

```
join_path(self.component_path, 'env', 'vars.sh'
```
Would be ```<prefix>/mkl/latest/env/vars.sh``` for mkl. For TBB, it would be ```<prefix>/tbb/latest/env/vars.sh``` and will set TBBROOT, etc. MPI sets I_MPI_HOME, among other things. The vars.sh also sets CPATH, LD_LIBRARY_PATH, LIBRARY_PATH. Some of these environment variables overlaps with the spack way of ensuring a build finds header files and libraries. Your suggestion of making mkl a link dependence seems like a better way to address the original problem.

There is still a problem that packages may assume that vars.sh has been sourced and look at HOME/ROOT type variables as part of the build. Sourcing vars.sh takes care of that, but I don't have a good feeling about whether this will create unwanted conflicts with spack mechanisms.

Reply:
@scheibelp Any updates on this issue?

Reply:
Hi @dmentock, based on https://github.com/spack/spack/pull/30802#discussion_r886675621

> Your suggestion of making mkl a link dependence seems like a better way to address the original problem.

I got the impression I would be seeing a new PR addressing this issue. If I misunderstood that (or you and @rscohn2 don't have the time to make that change) then I can revisit this or submit a PR to that effect.

Reply:
@dmentock : For mkl, I think setting the depends_on to be 'link' instead of 'build' would resolve the issue.  For the compiler, this PR resolves the issue. I suspect that https://github.com/spack/spack/pull/30981 would also work. This PR also takes care of setting the ROOT/HOME variables that the various vars.sh files set, or I could continue to set them manually on a package-by-package basis. 

@scheibelp, @skosukhin: We have 3 partially overlapping problems with 3 partially overlapping solutions. I want the set of fixes that will work best for spack users, but do not know how to sort it out.

If the spack team likes https://github.com/spack/spack/pull/30981 and it resolves @dmentock issue with the compiler, then I am ok with closing this PR and continuing to manually set the random HOME/ROOT variables in the package.py for each component.

Reply:
As far as I understand, #30981 solves a different problem: it makes `ifort -mkl` inject rpaths to the MKL libraries when called outside of Spack environment. In other words, the following should work but currently doesn't:

```console
$ module load intel-oneapi-compilers intel-oneapi-mkl
$ cat test.f90
program main
  print *, "All good!"
end program main
$ ifort -qmkl test.f90
$ ./a.out
 All good!
$ module unload intel-oneapi-compilers intel-oneapi-mkl
$ ./a.out
./a.out: error while loading shared libraries: libmkl_intel_lp64.so.2: cannot open shared object file: No such file or directory
```